DirectProgramming/DPC++FPGA/ReferenceDesigns/qrd/src/qrd.hpp

```diff
@@ -1,43 +1,153 @@
+                        // TT will be ac_complex<T> or T depending on is_complex
+         >
+void QRDecompositionImpl(
+  std::vector<TT> &a_matrix, // Input matrix to decompose
+  std::vector<TT> &q_matrix, // Output matrix Q
+  std::vector<TT> &r_matrix, // Output matrix R
+  sycl::queue &q,           // Device queue
+  size_t matrices,          // Number of matrices to process
+  size_t reps               // Number of repetitions, for performance evaluation
+) {
+
+  constexpr int kAMatrixSize = columns * rows;
+  constexpr int kQMatrixSize = columns * rows;
+  constexpr int kRMatrixSize = columns * (columns + 1) / 2;
+  constexpr int kNumElementsPerDDRBurst = is_complex ? 4 : 8;
+
+  using pipe_type = PipeTable<kNumElementsPerDDRBurst, TT>;
+
+  // Number of buffers to allocate to be able to read/compute/store without
+  // overlap.
+  constexpr short kNumBuffers = 3;
+
+  // Pipes to communicate the A, Q and R matrices between kernels
+  using a_matrix_pipe = sycl::ext::intel::pipe<class APipe, pipe_type, 3>;
+  using q_matrix_pipe = sycl::ext::intel::pipe<class QPipe, pipe_type, 3>;
+  using r_matrix_pipe = sycl::ext::intel::pipe<class RPipe, pipe_type, 3>;
+
+  // We will process 'matrices_per_iter' number of matrices in each run of the
+  // kernel
+#if defined(FPGA_EMULATOR)
+  constexpr int matrices_per_iter = 1;
+  assert(matrices % matrices_per_iter == 0);
+#else
+  constexpr int matrices_per_iter = 2048;
+  assert(matrices % matrices_per_iter == 0);
+#endif
+
+  // Create buffers and allocate space for them.
+  TT *a_device[kNumBuffers];
+  TT *q_device[kNumBuffers];
+  TT *r_device[kNumBuffers];
+  for (short i = 0; i < kNumBuffers; i++) {
+    a_device[i] = sycl::malloc_device<TT>(kAMatrixSize * matrices_per_iter, q);
+    q_device[i] = sycl::malloc_device<TT>(kAMatrixSize * matrices_per_iter, q);
+    r_device[i] = sycl::malloc_device<TT>(kRMatrixSize * matrices_per_iter, q);
+  }
+
+  // Repeat the computation multiple times (for performance analysis)
+  for (size_t r = 0; r < reps; r++) {
+    // Go over all the matrices, rotating buffers every time
+    for (size_t buffer_idx = 0, it = 0; it < matrices;
+         it += matrices_per_iter, buffer_idx = (buffer_idx + 1) % kNumBuffers) {
+      // Pointers to current input/output matrices in host memory
+      const TT *kPtrA = a_matrix.data() + kAMatrixSize * it;
+      TT *ptr_q = q_matrix.data() + kQMatrixSize * it;
+      TT *ptr_r = r_matrix.data() + kRMatrixSize * it;
+
+      auto copy_a_event = q.memcpy(a_device[buffer_idx], kPtrA,
+                                 kAMatrixSize * matrices_per_iter * sizeof(TT));
+
+      TT *current_a_buffer = a_device[buffer_idx];
+      TT *current_q_buffer = q_device[buffer_idx];
+      TT *current_r_buffer = r_device[buffer_idx];
+
+      auto read_event = q.submit([&](sycl::handler &h) {
+        h.depends_on(copy_a_event);
+        h.single_task<class QRDDDRToLocalMem>([=
+        ]() [[intel::kernel_args_restrict]] {
```

Comment:
Can this be on one line? I know clang-format messes it up, but it is nicer on one line IMO.